x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -197,15 +212,26 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
             searchResponse = null;
         } else {
             /*
+             * MP TODO: UPDATE THIS DOCUMENTATION TO MATCH NEW LOGIC
              * Build the response, reducing aggs if we haven't already and
              * storing the result of the reduction, so we won't have to reduce
              * the same aggregation results a second time if nothing has changed.
              * This does cost memory because we have a reference to the finally
              * reduced aggs sitting around which can't be GCed until we get an update.
              */
-            InternalAggregations reducedAggs = reducedAggsSource.get();
-            reducedAggsSource = () -> reducedAggs;
-            searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            if (searchResponseMerger == null) {
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            } else if (finalReduce == false) {
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                // TODO: check that this just returns the partial aggs response passed in if SearchResponseMerger
+                // has no other full SearchResponses
+                searchResponse = searchResponseMerger.getMergedResponse(clusters, buildResponse(task.getStartTimeNanos(), reducedAggs));
+            } else {
+                searchResponse = searchResponseMerger.getMergedResponse(clusters);
```

Comment:
should we assert that there are no partial aggs results in this scenario?

Reply:
No, there could be partial aggs results. We are just ignoring them on this branch because `localFinalReduce`=`true` (and CCS MRT=true). As described in the answer above, we can now ignore any partial results because the SearchResponseMerger has the (full/final) local cluster results.

Reply:
That sounds correct. I think that there could be scenarios where finalReduction has happened but the merger did not receive yet those local final results. If that's correct, do we just accept that we won't include such final results until the merger has been notified, hence there will be a little internal delay with CCS and MRT=true?

Reply:
Yes, I thought about that too. We'll need to decide if it is worth adding some sort additional logic to handle that slight gap.

*Considerations:*

We would have to add in some logic to determine whether or not the CCSActionListener for the local cluster has received the final SearchResponse (and added it to the SearchResponseMerger). There isn't anything in the SearchResponse passed to `SearchResponseMerger.add` that indicates what cluster it came from. So we could add a second `add` method (`addSearchResponseFromLocalCluster`) and then SearchResponseMerger could have a flag that is used by the MutableSearchResponse to determine whether to use the local partial aggs or not.

That would be straightforward, but again we are adding even more special handling for the local cluster in a cross-cluster search. We have one extreme (very clean codewise) where we don't do partial aggs for local cluster when CCS MRT=true -- all clusters are treated the same and the other where we have extra logic and tracking for the local cluster. Given that many/most large CCS customers don't have data on the local querying cluster, I'm in favor of the "cleaner" approach.

So options are:
1. don't do partial aggs for local when MRT=true
2. live with this slight gap in time such that in some cases the local cluster aggs could disappear from results we had in a previous response
3. add in additional logic to ensure that gap does not occur (tracking whether the local cluster has reported in via CCSActionListener/SearchResponseMerger).

Reply:
I have been thinking about alternatives: the root of the issue is that we introduced a separate path for minimize roundtrips that take local results from a different place, depending on a flag which is set to make the switch. But the flag says results are available in the merger before they actually are. I think finding a good solution for this is important. 

In short, I am thinking that instead of synchronizing, we should try to address the issue at the root, so that the problem we have to deal with no longer exists, if possible. One way to get there would be to go back to an additional callback in the progress listener, that is specific to getting results from remote clusters (and not the local one, as those come in via the existing callbacks). that can still use the search response merger under the hood to expose incremental results, but it would not rely on the same merger instance that is already created in TransportSearchAction, and it would not use the merger as an additional way to get notified of new response being made available. It would merely use it to merge response. 

I was considering ways to possibly using the merger for remote responses but not the local one, without going back to the additional progress listener callback, but I don't think that is possible.

Reply:
> > One way to get there would be to go back to an additional callback in the progress listener, that is specific to getting results from remote clusters (and not the local one, as those come in via the existing callbacks). 

What is this new callback sending? Is it sending individual `SearchResponse` objects? Is it sending merged SearchResponse objects like I had in an earlier iteration? Or only SearchResponse objects merged across all remote-only clusters?

What is the callback receiver (AsyncSearchTask/MutableSearchResponse) going to do with that callback object? Since it is being pushed when the clusters report back rather than pulled when the user is requesting an AsyncSearchResponse, we will have to cache it. But we discussed in my earlier iteration that we don't want to cache the newly merged SearchResponses to avoid using more memory.

If the callback receiver is receiving individual per-cluster SearchResponse objects, then we are double caching those between the two SearchResponseMergers you propose. At that point I don't see what the point of the TransportSearchAction-based SearchResponseMerger is. It is redundant.

> that can still use the search response merger under the hood to expose incremental results, but it would not rely on the same merger instance that is already created in TransportSearchAction, and it would not use the merger as an additional way to get notified of new response being made available. It would merely use it to merge response.

Same comment as above. I don't see how two SearchResponseMergers is better, only worse. The problem becomes how does the 2nd SearchResponseMerger get access to the finished responses without using additional RAM (caching)?

Reply:
I have a new idea and a new commit that I like (I hope you do too!): https://github.com/elastic/elasticsearch/pull/103134/commits/289955a6e1b3aa59c6a1f676edc7ade0a691f049

I think this is a pretty clean solution. It does require the `SearchResponseMerger` to become aware that it may be passed a SearchResponse from the local cluster that is a temporary partial response. 

The MutableSearchResponse object can ask the SearchResponseMerger: do you have a result from the local cluster? If not, it creates a SearchResponse from the local partial aggs and passes that in. But there's a race condition (that I didn't handle before) where we might have local results twice, so the `SearchResponseMerger.getMergedResponse` method now checks whether the final local SearchResponse has come in and if so, doesn't use the local partial aggs response.

To my thinking, this expansion of awareness is not out of place - the SearchResponseMerger is only used for MRT=true and its job is to merge responses. It now participates in doing partial result merges safely.

Reply:
I'd like to better understand the implications around an additional callback in the progress listener. The problem we used to have with memory usage was around caching the finally reduced response (what getMergedResponse returns), but we are no longer doing that. Getting notified of a new search response that came in, that the merger needs to track too, only adds a new reference to an object that is already held in memory? The important bit is that we don't end up adding new instances of SearchResponses to the heap, which I don't think we would do. What am I missing?

The atomic boolean is ok, but I think we are adding complexity to the search response merger and making it become a component that async search relies on, not only for merging responses but to get somehow notified of new results. I don't like that, because it introduces weakness and complexity around needing to monitor results from two different places. I think it would be much cleaner to have all the logic about the progress of async search in one place. Also, given you need to introduce the distinction between local and remote results in the merger, that I don't think belongs there, it feels like remote results should have their own callback in the progress listener, as this would clarify the contract and make things easier to follow.

First though, let's clarify the memory issue, because indeed that is the most important aspect and why we have moved to the current approach.

Reply:
> let's clarify the memory issue,

You are correct that having MutableSearchResponse/AsyncSearchTask hold onto the same SearchResponses as SearchResponseMerger does not increase memory (where a newly merged SearchResponse would). However, ti does require additional ref-counting. It also potentially opens the question from maintainers of why are we using two different SearchResponseMergers which look somewhat redundant.

Taking your feedback, I've pushed two new commits as next attempts.

**Latest code changes**

I reverted the previous commit where SearchResponseMerger held logic to decide whether to use the localPartialAggs or the local SearchResponse.

The [first new commit](https://github.com/elastic/elasticsearch/pull/103134/commits/7dea93a54516de9df09a7f4caf443f698a4cd3e7) adds in a SearchProgressListener callback so that MutableSearchResponse can track accurately whether the SearchResponse for the local cluster has been added to the (single) SearchResponseMerger. In this one, the MutableSearchResponse does not have a second SearchResponseMerger. The problem with this commit is that it reintroduces the race condition between using local-partial-aggs vs. local full results and could end up using both. 

The [second new commit](https://github.com/elastic/elasticsearch/pull/103134/commits/298e533d401214d955aa0328ec643e5b35a50e1e) changes the new SearchProgressListener callback to also include the SearchResponse from each cluster, so those can be collected by AsyncSearchTask/MutableSearchResponse, which will create their own SearchResponseMerger and solve the race condition above without having to add new logic to SearchResponseMerger.

There are several different ways to structure this. I picked one that seemed like the least amount of code changes and has consistency guarantees (e.g., avoids the race condition).

The problems to solve are:

• how do you provide the constructor arguments to SearchResponseMerger, many of which are created in the TransportSearchAction

• who should collect and hold the SearchResponses passed into the new SearchProgressListener - AsyncSearchTask, MutableSearchResponse, or maybe some new PartialResponseHandler class that they delegate to?

• should MutableSearchResponse create a new SearchResponseMerger on the fly or create a single one that is reused? 

If we create a single one that is reused then it has to be updated as each SearchResponse comes into the SearchProgressListener, so maybe it should be the thing that holds the cluster SearchResponses? But that feels pretty redundant to the other SearchResponseMerger held by the TransportSearchAction/CCSActionListener's. And (unless we apply synchronization) it doesn't solve the race condition between having partial local aggs and local full results.

• where should logic for how to create partial results reside? In my latest commit, I've left it in the "partial results branch" of MutableSearchResponse.toAsyncSearchResponse, but perhaps we should create some PartialResponseHandler as I suggested above to handle that.


Also: I can't merge in `main` any longer, as Armin has added ref-counting to this part of the code and I'll have merge conflicts to deal with, so I'd rather put that off until we agree on a final design for this PR.

Reply:
I am not too concerned with questions around using two different search response mergers. I am more concerned about using the merger to signal that new responses have come in, which is difficult to reason about because there are then two ways that results may come from.

I am happy with the additional callback you added and the mechanism you built to decide when partial aggs have to be taken into account or not. Perhaps in the future this will change once we expose partial top hits for local results too.

As for the problems that you listed for solving, I don't have strong preferences on either of them, and I'd leave it up to you to polish this and finish it up, unless there are challenges about picking one approach over the other.

Thanks!